src/fork_gc.c

```diff
@@ -322,177 +344,179 @@ static void ForkGc_CollectGarbageFromTagIdx(ForkGCCtx *gc, RedisSearchCtx *sctx)
+  blockModified->blockOldIndex = msg.oldix;
+  blockModified->blk.firstId = msg.firstId;
+  blockModified->blk.lastId = msg.lastId;
+  blockModified->blk.numDocs = msg.numDocs;
 
   Buffer *b = &blockModified->blk.buf;
-  b->data = ForkGc_FDReadBuffer(gc->pipefd[GC_READERFD], &b->offset);
+  b->data = FGC_RecvBuffer(gc, &b->offset);
   b->cap = b->offset;
 }
 
-static bool ForkGc_ReadInvertedIndexFromFork(ForkGCCtx *gc, ForkGc_InvertedIndexData *idxData) {
-  long long blocksRepaired = ForkGc_FDReadLongLong(gc->pipefd[GC_READERFD]);
-  if (!blocksRepaired) {
+static bool FGC_RecvInvIdx(ForkGCCtx *gc, InvIdxInfo *info) {
+  long long wasRepaired = FGC_RecvLongLong(gc);
+  if (!wasRepaired) {
     return false;
   }
-
-  idxData->originalSize = ForkGc_FDReadLongLong(gc->pipefd[GC_READERFD]);
-
-  idxData->blockArrayExists = ForkGc_FDReadLongLong(gc->pipefd[GC_READERFD]);
-  if (idxData->blockArrayExists) {
-    idxData->newBlocksArray =
-        (IndexBlock *)ForkGc_FDReadBuffer(gc->pipefd[GC_READERFD], &idxData->newBlocksArraySize);
-    idxData->newBlocksArraySize /= sizeof(IndexBlock);
+  FGC_MsgInvIdxInfo infmsg = {0};
+  FGC_RecvFixed(gc, &infmsg, sizeof infmsg);
+  info->newBlocks = FGC_RecvBuffer(gc, &info->numNewBlocks);
+  if (info->numNewBlocks) {
+    info->numNewBlocks /= sizeof(*info->newBlocks);
   }
 
-  idxData->freeBufs = (char **)ForkGc_FDReadBuffer(gc->pipefd[GC_READERFD], &idxData->freeBufsSize);
-  idxData->freeBufsSize /= sizeof(char *);
+  info->addrsToFree = FGC_RecvBuffer(gc, &info->numAddrsToFree);
+  info->numAddrsToFree /= sizeof(char *);
 
-  long long blocksModifiedSize = ForkGc_FDReadLongLong(gc->pipefd[GC_READERFD]);
+  info->originalSize = infmsg.nblocksOrig;
+  info->numChangedBlocks = infmsg.nblocksRepaired;
+  info->bytesCollected = infmsg.nbytesCollected;
+  info->docsCollected = infmsg.ndocsCollected;
 
-  idxData->bytesCollected = ForkGc_FDReadLongLong(gc->pipefd[GC_READERFD]);
-  idxData->docsCollected = ForkGc_FDReadLongLong(gc->pipefd[GC_READERFD]);
-
-  idxData->blocksModified = array_new(ModifiedBlock, blocksModifiedSize);
-  for (int i = 0; i < blocksModifiedSize; ++i) {
-    ModifiedBlock mb;
-    ForkGc_ReadModifiedBlock(gc, &mb);
-    idxData->blocksModified = array_append(idxData->blocksModified, mb);
+  info->changedBlocks = rm_malloc(sizeof(*info->changedBlocks) * info->numChangedBlocks);
+  for (size_t i = 0; i < info->numChangedBlocks; ++i) {
+    FGC_RecvModifiedBlock(gc, info->changedBlocks + i);
   }
   return true;
 }
 
-static void ForkGc_FixInvertedIndex(ForkGCCtx *gc, ForkGc_InvertedIndexData *idxData,
-                                    InvertedIndex *idx) {
-  if (idxData->freeBufs) {
-    for (int i = 0; i < idxData->freeBufsSize; ++i) {
-      rm_free(idxData->freeBufs[i]);
-    }
-    rm_free(idxData->freeBufs);
+static void ForkGc_FixInvertedIndex(ForkGCCtx *gc, InvIdxInfo *idxData, InvertedIndex *idx) {
+  for (size_t i = 0; i < idxData->numAddrsToFree; ++i) {
```

Comment:
What if the last index is in this list? it will be free while actually there is a possibility that new docs was added to it? Even worst it might got realloc and you will free an already freed memory.

Reply:
Good catch